src/main/java/org/spongepowered/api/service/permission/SubjectCollection.java

```diff
@@ -42,54 +58,197 @@
 
     /**
-     * Returns all subjects. The iterator provided by this method may be
-     * populated asynchronously. Be aware that for large collections this method
-     * may be very resource-intensive
+     * Performs an action on each Subject in the collection.
      *
-     * @return An iterable providing all subjects stored by this collection.
+     * <p>Subjects are loaded, supplied to the consumer, and then allowed to be
+     * uncached by the implementation.</p>
+     *
+     * <p>This should be used to apply bulk changes or gather data about all
+     * Subjects in the collection. The provided consumer will be supplied
+     * asynchronously. Acting upon a large collection may be particularly
+     * resource intensive.</p>
+     *
+     * <p>Implementations may choose to load subjects in parallel or
+     * sequentially.</p>
+     *
+     * @param action The action to perform on each subject
+     * @return A future which will complete when the operation has finished
      */
-    Iterable<Subject> getAllSubjects();
+    default CompletableFuture<Void> applyToAll(Consumer<Subject> action) {
+        Preconditions.checkNotNull(action, "action");
+        return CompletableFuture.runAsync(() -> {
+            Set<String> identifiers = getAllIdentifiers().join();
+            applyToAll(action, identifiers);
+        });
+    }
 
     /**
-     * Return all known subjects with the given permission information. Because
-     * no context is passed, only subjects who have this permission globally or
-     * subjects which have accurate context calculations are returned.
+     * Performs an action on each Subject in the provided set.
+     *
+     * <p>Subjects are loaded, supplied to the consumer, and then allowed to be
+     * uncached by the implementation.</p>
+     *
+     * <p>This should be used to apply bulk changes or gather data about all
+     * Subjects in the collection. The provided consumer will be supplied
+     * asynchronously. Acting upon a large collection may be particularly
+     * resource intensive.</p>
+     *
+     * <p>Implementations may choose to load subjects in parallel or
+     * sequentially.</p>
+     *
+     * @param action The action to perform on each subject
+     * @param identifiers a set of identifiers to apply the action to
+     * @return A future which will complete when the operation has finished
+     */
+    default CompletableFuture<Void> applyToAll(Consumer<Subject> action, Set<String> identifiers) {
+        Preconditions.checkNotNull(action, "action");
+        Preconditions.checkNotNull(identifiers, "identifiers");
+        return CompletableFuture.runAsync(() -> {
+            for (String id : identifiers) {
+                try {
+                    Subject s = loadSubject(id).get();
+                    action.accept(s);
+                    suggestUnload(s.getIdentifier());
+                } catch (Exception e) {
+                    e.printStackTrace();
```

Comment:
I'm not sure whether it is wise to just print the stack to the console. It would probably be better to log to slf4j or a Logging framework or at least cancel the execution. I would be fine if you keep a reference to the first exception that happened and add the others as suppressed and after the invocation propagate that exception up.